### ISSUE TITLE
Scaled loss in Mimo

### ISSUE THREAD

**Issue by alexrame** [google/edward2#428]
Thanks for the really great work.
I really like your approach in Mimo, I hope you will get it into ICLR.
I am trying to reproduce your results on one single GPU (and in Pytorch) for computational reasons but still can not find the good hyperparameters.

I guess my problem is related to the scaling of your loss on several cores. Indeed, in your implementation:
```
# Scale the loss given the TPUStrategy will reduce sum all gradients.
loss = negative_log_likelihood + l2_loss
scaled_loss = loss / strategy.num_replicas_in_sync
```
I do not understand why you divide your loss by  num_replicas_in_sync (equal to 8 in your case). It seems like you are averaging negative_log_likelihood, that has already been previously averaged.

Intuitively, I would have done:
```
scaled_loss = (negative_log_likelihood + l2_loss)*strategy.num_replicas_in_sync
```
or maybe
```
scaled_loss = negative_log_likelihood*strategy.num_replicas_in_sync  + l2_loss
```

Could you explain the reasons behind your choice on this "scaled_loss"? 
Thanks in advance, and good luck with your future projects.

**Comment by mhavasi** [comment]
Hi, sorry for the late reply.

The negative_log_likelihood is only averaged within the TPU's minibatch. So if the per_core_batch_size is 64, then this is only the average of 64 inputs. To get the correct gradients, we need to divide the loss by the number of replicas because the gradients from the replicas are summed at the end.

If we didn't divide by the number of gradients, each TPU would return the gradients for its own minibatch, which we would then sum up and get num_replicas_in_sync times the gradient of the overall loss.

**Comment by alexrame** [comment]
Thank you for your time and help.
And by the way, I was able to reproduce (and even improve) most of MIMO results on a single GPU in Pytorch.
You may find the detailed performance in our recent preprint: https://arxiv.org/abs/2103.06132
I will release my code soon.
I hope to catch you at iclr :)
